src/agent/block_store_services/block_store_client.js

```diff
@@ -43,223 +60,354 @@ class BlockStoreClient {
                 return this._delegate_read_block_s3(rpc_client, params, options);
             case 'BLOCK_STORE_AZURE':
                 return this._delegate_read_block_azure(rpc_client, params, options);
+            case 'BLOCK_STORE_GOOGLE':
+                return this._delegate_read_block_google(rpc_client, params, options);
             default:
                 return rpc_client.block_store.read_block(params, options);
         }
     }
 
-    _delegate_write_block_azure(rpc_client, params, options) {
+    _write_to_stream(writable, buf) {
+        return new P((resolve, reject) => {
+            writable.once('error', reject);
+            writable.write(buf, err => {
+                if (err) {
+                    return reject(err);
+                }
+                return resolve();
+            });
+        });
+    }
+
+    async _delegate_write_block_google(rpc_client, params, options) {
         const { timeout = config.IO_WRITE_BLOCK_TIMEOUT } = options;
         const { block_md } = params;
         const data = params[RPC_BUFFERS].data;
-        return rpc_client.block_store.delegate_write_block({ block_md, data_length: data.length }, options)
-            .catch(err => {
-                dbg.error('failed to get signed access information from cloud agent', err);
-                // assuming that usage was not updated yet, so no need to update usage after error
-                throw err;
-            })
-            .then(delegation_info => {
-                const { host, container, block_key, blob_sas, metadata, usage } = delegation_info;
-                // create a shared blob service using the blob_sas (shared access signature)
-                const shared_blob_svc = azure_storage.createBlobServiceWithSas(host, blob_sas);
-                return P.fromCallback(callback => shared_blob_svc.createBlockBlobFromText(container,
-                        block_key,
-                        data, { metadata },
-                        callback))
-                    .catch(err => {
-                        dbg.error('encountered error on _delegate_write_block_azure:', err);
-                        const error = _.pick(err, 'message', 'code');
-                        return rpc_client.block_store.handle_delegator_error({ error, usage, op_type: 'WRITE' }, options);
-                    });
-            })
-            .timeout(timeout);
+        let bs_info;
+        let bucket;
+        return promise_utils.timeout(async () => {
+            try {
+                bs_info = await block_store_info_cache.get_with_cache({ options, rpc_client });
+                if (!bs_info) throw new Error('couldn\'t resolve cloud credentials');
+                const google_params = bs_info.connection_params;
+                const block_id = block_md.id;
+                const block_dir = get_block_internal_dir(block_id);
+                const google = new Storage({
```

Comment:
Best move this to occur once in the cache load

Reply:
didn't change for now. on next PR.